fetch-in-react-practice/client/components/app.jsx

```diff
@@ -0,0 +1,111 @@
+     * update state with the received todos.
+     */
+    fetch('/api/todos', { method: 'GET' })
+      .then(response => response.json())
+      .then(todoList => {
+        this.setState({ todos: todoList });
+      })
+      .catch(err => console.error('Fetch failed', err));
+  }
+
+  addTodo(newTodo) {
+    /**
+    * Use fetch to send a POST request to `/api/todos`.
+    * Then 😉, once the response JSON is received and parsed,
+    * add the created todo to the state array.
+    *
+    * Do not mutate the original state array, nor any objects within it.
+    * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+    *
+    * TIP: Be sure to SERIALIZE the todo object in the body with JSON.stringify()
+    * and specify the "Content-Type" header as "application/json"
+    *
+    * TIP: Use Array.prototype.concat to create a new array containing the contents
+    * of the old array, plus the object returned by the server.
+    */
+    fetch('/api/todos', {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newTodo)
+    })
+      .then(response => response.json())
+      .then(newTodo => {
+        const updatedList = this.state.todos.concat(newTodo);
+        this.setState({ todos: updatedList });
+      })
+      .catch(err => console.error('Fetch failed', err));
+
+  }
+
+  toggleCompleted(todoId) {
+    /**
+     * Find the index of the todo with the matching todoId in the state array.
+     * Get its "isCompleted" status.
+     * Make a new object containing the opposite "isCompleted" status.
+     * Use fetch to send a PATCH request to `/api/todos/${todoId}`
+     * Then 😉, once the response JSON is received and parsed,
+     * replace the old todo in the state array.
+     *
+     * NOTE: "toggle" means to flip back and forth, so clicking a todo
+     * in the list should "toggle" its isCompleted status back and forth.
+     *
+     * Do not mutate the original state array, nor any objects within it.
+     * https://reactjs.org/docs/optimizing-performance.html#the-power-of-not-mutating-data
+     *
+     * TIP: Be sure to SERIALIZE the updates in the body with JSON.stringify()
+     * And specify the "Content-Type" header as "application/json"
+     */
+
+    const index = this.state.todos.findIndex(item => item.todoId === todoId);
+    const clickedItem = this.state.todos[index];
+    const completedStatus = clickedItem.isCompleted;
+    const newCompletedStatus = { isCompleted: !completedStatus };
+
+    fetch(`/api/todos/${todoId}`, {
+      method: 'PATCH',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(newCompletedStatus)
+    })
+      .then(response => response.json())
+      .then(newStatus => {
+        const oldList = this.state.todos;
+        const updatedStatusList = Object.assign(oldList, newStatus);
+        this.setState({ todos: updatedStatusList });
+        this.componentDidMount();
```

Comment:
We never want to directly call one of the lifecycle methods. If state is updated properly you wont need this. You can use the `map` method to build a new todos array from the array in state and replace the correct todo item.

```JAVASCRIPT
.then(updated => {
  const allTodos = this.state.todos.map(original => {
    return original.todoId === updated.todoId
      ? updated
      : original;
  });
  this.setState({ todos: allTodos });
});
```

Reply:
Sounds good, thanks!